r-package/grf/R/boosted_regression_forest.R

```diff
@@ -0,0 +1,229 @@
+#' @examples \dontrun{
+#' # Train a boosted regression forest.
+#' n = 50; p = 10
+#' X = matrix(rnorm(n*p), n, p)
+#' Y = X[,1] * rnorm(n)
+#' r.b.forest = boosted_regression_forest(X, Y)
+#'
+#' # Predict using the forest.
+#' X.test = matrix(0, 101, p)
+#' X.test[,1] = seq(-2, 2, length.out = 101)
+#' r.pred = predict(r.b.forest, X.test)
+#'
+#' # Predict on out-of-bag training samples.
+#' r.pred = predict(r.b.forest)
+#'
+#' #Check how many boosting iterations were used
+#' print(length(r.b.forest$forests))
+#' }
+#'
+#' @export
+boosted_regression_forest <- function(X, Y,
+                                      sample.fraction = 0.5,
+                                      mtry = NULL,
+                                      num.trees = 500,
+                                      num.threads = NULL,
+                                      min.node.size = NULL,
+                                      honesty = TRUE,
+                                      honesty.fraction = NULL,
+                                      ci.group.size = 2,
+                                      alpha = NULL,
+                                      imbalance.penalty = NULL,
+                                      seed = NULL,
+                                      clusters = NULL,
+                                      samples_per_cluster = NULL,
+                                      tune.parameters = FALSE,
+                                      num.fit.trees = 10,
+                                      num.fit.reps = 100,
+                                      num.optimize.reps = 1000,
+                                      num.steps = NULL,
+                                      tolerance = 0.95,
+                                      max.steps = 6,
+                                      num.trees.tune = 10) {
+
+  boosted.forest = NULL
+  boosted.forest[["forests"]] = list()
+  boosted.forest[["debiased.errors"]] = list()
+  forest.Y <- regression_forest(X, Y, sample.fraction = sample.fraction,
+                                mtry = mtry, tune.parameters = tune.parameters,
+                                num.trees = num.trees,
+                                num.threads = num.threads,
+                                min.node.size = min.node.size, honesty = honesty,
+                                honesty.fraction = honesty.fraction,
+                                seed = seed, ci.group.size = ci.group.size,
+                                alpha = alpha,
+                                imbalance.penalty = imbalance.penalty,
+                                clusters = clusters, samples_per_cluster = samples_per_cluster);
+  current.pred <- predict(forest.Y)
+  #save tuned parameters for use on future boosting iterations
+  tunable.params <- forest.Y$tunable.params
+  Y.hat <- current.pred$predictions
+  error.debiased <- current.pred$debiased.error
+  boosted.forest[["forests"]][[1]] <- forest.Y
+  boosted.forest[["debiased.errors"]][[1]] <- error.debiased
```

Comment:
I think we should store the mean debiased error (and call it `error`) instead of the pointwise errors. This is more consistent with what we return from other tuning methods like `regression_tuning`.